src/python/pants/engine/target.py

```diff
@@ -1387,7 +1361,109 @@ async def hydrate_sources(
         )
     )
     sources_field.validate_snapshot(snapshot)
-    return HydratedSources(snapshot, sources_field.filespec)
+
+    should_generate_sources = request.codegen_language is not None and isinstance(
+        sources_field, CodegenSources
+    )
+    if not should_generate_sources:
+        return HydratedSources(snapshot, sources_field.filespec)
+
+    generate_request_types: Iterable[Type[GenerateSourcesRequest]] = union_membership.union_rules[
+        GenerateSourcesRequest
+    ]
+    relevant_generate_request_types = [
+        generate_request_type
+        for generate_request_type in generate_request_types
+        if isinstance(sources_field, generate_request_type.input)
+        and generate_request_type.output == request.codegen_language
+    ]
+    if not relevant_generate_request_types:
+        return HydratedSources(EMPTY_SNAPSHOT, sources_field.filespec)
```

Comment:
Alternatively, we could error when the codegen is not possible, but no-oping is much, much friendlier to do and makes call-sites substantially more ergonomic. The caller should not need to care whether a particular codegen implementation exists; instead, they're saying "Try your best to get my Python, if possible."

Another approach is to return `Optional[HydratedSources]`, but I'd like to avoid inventing new engine features here, if possible.

Reply:
Using this "null object" pattern should be fine for now I think.

Reply:
> The reason for this is that it's possible that other fields of the object indicate that something is a legal target (for example), but that having empty sources would not be legal.

This is sort of handled today by `Sources` subclasses being able to a) mark the `sources` field as required, and b) setting `expected_num_files` to validate, for example, that it's > 0. 

Why I say sort of is that this only allows target authors to express the requiredness, rather than rule call sites.

--

Generally, I agree it's more sound to not use the null-object pattern with `EMPTY_SNAPSHOT`.

But, I'm a little concerned to make the change because it makes call sites much less ergonomic. 
Now, to see if a resulting `HydratedSources` has files:

```python
if hydrated_sources.snapshot and hydrated_sources.snapshot.files`
```

To use the resulting file:

```python
file_names = hydrated_sources.snapshot.files if hydrated_sources.snapshot else ()
```

If you still think we should make this change, I'll put up a dedicated PR to make the field be `Optional` so that we can better evaluate if it's a worthwhile change or not.

Reply:
Regarding making the inner value optional: I think I had an interesting thought in https://github.com/pantsbuild/pants/pull/9634#discussion_r416121567

Basically: I think that it's possible that the SourcesRequest could handle _all_ of the casting in one go... so then it would basically be: "get me this target as either an X, Y, or Z"... then the caller would switch on the possibilities.

The advantage to this is that you almost ~never want to treat a field as multiple types in one consumer (ie, it's very unlikely for me to want a dependency as both Files and PythonSources), so the caller will want to do something like: `"for each dependency, get as exactly one of these field types, or None if I should ignore it"`.